README
Parallel-SubFork
================

This module provides a simple way to execute arbitrary perl code from different
processes through the fork system call.

INSTALLATION

To install this module type the following:

	perl Makefile.PL
	make
	make test
	make install

DEPENDENCIES

This module depends on the following CPAN modules:

	Carp
	POSIX
	Class::Accessor::Fast

COPYRIGHT AND LICENCE

Copyright (C) 2008 by Emmanuel Rodriguez

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.


Changes
Revision history for Perl extension Parallel::SubFork.

0.09 Sat Jan 16 16:37:15 CET 2010
	- Documentation update.

0.08 Sat Nov 14 13:02:05 CET 2009
	- Implement the timeout in $task->wait_for through sleep instead of alarm.
	  [Ticket #51465]

0.07 Thu Nov 12 20:59:10 CET 2009
	- Add a prototype to sub_fork.

0.06 Thu May  7 07:42:05 CEST 2009
	- New release with all the changes from the 0.06_XX releases.

0.06_02  Fri Feb  6 22:34:52 CET 2009 
	- Added unit tests for wait_for($timeout).
	- The unit tests are skipped if the semaphore can't be created.

0.06_01  Fri Feb  5 19:56:21 CET 2009 
	- The method wait_for() can accept a timeout.

0.05  Sun Mar 16 09:57:14 2008 
	- Added the function sub_fork.
	- Updated the documentation: removed all links to the functions as CPAN can't
	  link to them properly.

0.04  Wed Mar 12 18:37:41 2008 
	- The tests where missing the file Tsemaphore.pm
	  [Ticket #34016]

0.03  Tue Mar 11 21:27:48 2008 
	- Fixed a race condition in the tests by replacing kill/pause with semaphores
	  [Ticket #33955]
	- Skipping the test under Windows since fork is broken and unreliable and 
	  IPC::SysV is missing

0.02  Mon Mar  10 07:33:57 2008
	- Fixed the AUTHOR entry in Makefile.PL
	- Formatted some white spaces in the LICENSE

0.01  Sat Mar  8 09:13:14 2008
	- original version; created by h2xs 1.23 with options
		-XA -n Parallel::SubFork


t/Tsemaphore.pm
#!/usr/bin/perl
package Tsemaphore;


use strict;
use warnings;

use IPC::SysV qw(IPC_PRIVATE S_IRWXU IPC_CREAT);
use IPC::Semaphore;
use POSIX qw(WNOHANG);

use Test::More;

# This is a testing framework so we can do ugly things like exporting symbols by
# default into the caller's context
use base 'Exporter';
our @EXPORT = qw(
	semaphore_init
	semaphore_reset
	semaphore_task
	test_semaphore_task_run
);

#	semaphore_let_go
#	semaphore_wait_for
#	$SEMAPHORE_POINT_A
#	$SEMAPHORE_POINT_B

# Semaphores are used for synchronizing the parent (main code) and the child
# (task). The idea is that the parent will check that the child is actually
# running. In order for this test to be successful the child has to be running!
# To ensure that the child is running semaphores are used. The child will not be
# allowed to finish until the parent has approved it. On the other hand the 
# parent will not be allowed to check for child's process until the child is
# alive and notifies the parent of his existence.
#
# Parent                 |  Child
# -----------------------|-----------
#                        |
# Init                   |
# Create semaphores      |
#                        |
# Fork task              | Task start
#                        |
# Wait for A             | Join A
#                        |
# Check child is running |
#                        |
# Join B                 | Wait for B
#                        |
# Continue tests         | Finish
#                        |
# Finish                 |
#
my $SEMAPHORE;
my $SEMAPHORE_POINT_A = 0;
my $SEMAPHORE_POINT_B = 1;

my $PID = $$;

END {
	$SEMAPHORE->remove if defined $SEMAPHORE;
}

#
# Creates a new set of semaphores. If the semaphores can't be created it returns
# false.
#
sub semaphore_init {
	my ($skip_count) = @_;
	die "Usage: semaphore_init(skip_count)" unless @_;
	
	# Remove the previous semaphore
	$SEMAPHORE->remove if defined $SEMAPHORE;
	
	# Create a semaphore holding 2 values
	$SEMAPHORE = IPC::Semaphore->new(IPC_PRIVATE, 2, S_IRWXU | IPC_CREAT);
	if (! defined $SEMAPHORE) {
		# Bad implementation of IPC::Semaphore
		SKIP: {
			skip "Can't create a semaphore", $skip_count;
		}
		return 0;
	}
	isa_ok($SEMAPHORE, 'IPC::Semaphore');
	
	semaphore_reset();
	return 1;
}


#
# Resets the semaphores to 0
#
sub semaphore_reset {
	# Clear the semaphores
	my $return = $SEMAPHORE->setall(0, 0);
	ok(defined($return), "Semaphore cleared");
}


#
# Tell the other process that he can go futher since we have reached the rally
# point. We give the other process one more resource in order to go on.
#
sub semaphore_let_go {
	my ($who) = @_;
	$SEMAPHORE->op($who, 1, 0);
}


#
# Wait for the other process to reach his rally point and to let us go further.
# We remove a resource from this process, this will make us wait until the other
# process reaches the rally point.
#
sub semaphore_wait_for {
	my ($who) = @_;
	$SEMAPHORE->op($who, -1, 0);
}


#
# Special task that's meant to be run in a separate process. The return value is 
# used to tell if the task run successfully. If this task return 57 any other
# value must be regarded as a failure.
#
sub semaphore_task {
	my (@args) = @_;
	
	# Make sure that there's no hanging
	alarm(10);
	

	# Tell the parent that we are ready
	semaphore_let_go($SEMAPHORE_POINT_A) or return 10;

	
	# Wait for the parent to let us go further
	semaphore_wait_for($SEMAPHORE_POINT_B) or return 11;

	return 12 unless $$ != $PID;
	
	my @wanted = qw(1 2 3 4 5 6 7 8 9 10);
	return 13 unless eq_array(\@args, \@wanted);
	
	return 57;
}


#
# Execute a task and test that it's running properly. If a callback is passed
# then it will be invoked while the task is running.
#
sub test_semaphore_task_run {
	my ($task, $callback) = @_;

	# Make sure that there's no hanging, it's better to fail the test due to a
	# timeout than to leave the test haging there forever.
	alarm(10);
	

	isa_ok($task, 'Parallel::SubFork::Task');
	
	# Wait for the kid to be ready
	my $return = semaphore_wait_for($SEMAPHORE_POINT_A);


	# Make sure that the task is in a different process
	ok($$ != $task->pid, "Taks has a different PID");
	{
		my $kid = waitpid($task->pid, WNOHANG);
		is($kid, 0, "Child process still running");
	}
	
	
	# If the user provided a callback invoke it.
	if (ref $callback eq 'CODE') {
		$callback->($task);
	}
	

	# Tell the kid that we finish checking it, it can now resume
	$return = semaphore_let_go($SEMAPHORE_POINT_B);
	ok($return, "Removed resource to semaphore B");
	
	
	
	# Wait for the task to resume
	$task->wait_for();
	is($task->exit_code, 57, "Task exit code is fine");
	is($task->status, 57 << 8, "Task status is fine");

	is_deeply(
		[ $task->args ], 
		[ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 ], 
		"Task args are intact"
	);

	# Wait some more, is useless but it should work
	$task->wait_for();
	is($task->exit_code, 57, "Second wait on the same task, exit code fine");
	is($task->status, 57 << 8, "Second wait on the same task, status fine");
	

	# Make sure that there are no other tasks
	{
		my $kid = waitpid(-1, WNOHANG);
		is($kid, -1, "No more processes");
	}
}


# Return a true value
1;



lib/Parallel/SubFork.pm
package Parallel::SubFork;

=head1 NAME

Parallel::SubFork - Manage Perl functions in forked processes. 

=head1 SYNOPSIS

	use Parallel::SubFork;
	my $manager = Parallel::SubFork->new();
	
	# Start two parallel tasks
	$manager->start(sub { sleep 10; print "Done\n" });
	$manager->start(\&callback, @args);
	
	# Wait for all tasks to resume
	$manager->wait_for_all();
	
	# Loop through all tasks
	foreach my $task ($manager->tasks) {
		# Access any of the properties
		printf "Task with PID %d resumed\n", $task->pid;
		printf "Exist status: %d, exit code: %d\n", $task->status, $task->exit_code;
		printf "Args of task where: %s\n", join(', ', $task->args);
		print "\n";
	}

or more easily:

	use Parallel::SubFork qw(sub_fork);
	
	my $task = sub_fork(\&callback, @args);
	$task->wait_for();

=head1 DESCRIPTION

This module provides a simple wrapper over the module L<Parallel::SubFork::Task>
which in turns simplifies the usage of the system calls C<fork> and C<waitpid>.
The idea is to isolate the tasks to be execute in functions or closures and to
execute them in a separated process in order to take advantage of
parallelization.

=head1 TASKS

A task is simply a Perl function or a closure that will get executed in a
different process. This module will take care of creating and managing the new
processes. All that's left is to code the logic of each task and to provide the
proper I<inter process communication> (IPC) mechanism if needed.

A task will run in it's own process thus it's important to understand that all
modifications to variables within the function, even global variables, will have
no impact on the parent process. Communication or data exchange between the task
and the dispatcher (the code that started the task) has to be performed through
standard IPC mechanisms. For further details on how to establish different
communication channels refer to the documentation of L<perlipc>.

Since a task is running within a process it's expected that the task will return
an exit code (C<0> for an execution without flaws and any other integer for
reporting an error) and not a true value in the I<Perl> sense. The return value
will be used as the exit code of the process that's running the task.

=cut

use strict;
use warnings;

use Carp;

use base 'Exporter';
our @EXPORT_OK = qw(
	sub_fork
);

use Parallel::SubFork::Task;

use base qw(Class::Accessor::Fast);
__PACKAGE__->mk_accessors(
	qw(
		tasks
		_dispatcher_pid
	)
);


# Version of the module
our $VERSION = '0.09';


=head1 FUNCTIONS

The module provides the following functions:

=cut


=head2 sub_fork

This function provides a simple way for creating and launching tasks. It is
declared using a prototype which allows it to be called as:

	my $task = sub_fork { print "$$ > $_\n" for 1 .. 10 };
	$task->wait_for();

Parameters:

=over

=item $code

The code reference to execute.

=item @args (optional)

The arguments to pass to the code reference.

=back

=cut

sub sub_fork (&;@) {

	# Arguments
	my ($code, @args) = @_;

	my $task;
	eval {
		$task = Parallel::SubFork::Task->start($code, @args);
	};
	if (my $error = $@) {
		croak $error;
	}
	return $task;
}


=head1 METHODS

The module defines the following methods:

=cut


=head2 new

Creates a new C<Parallel::SubFork>.

=cut

sub new {

	# Arguments
	my $class = shift;
	
	# Create a blessed instance
	my $self = bless {}, ref($class) || $class;
	
	# The list of children spawned
	$self->tasks([]);
	
	# The PID of the dispacher
	$self->_dispatcher_pid($$);

	return $self;
}


=head2 start

Starts the execution of a new task in a different process. A task consists of a
code reference (a closure or a reference to a subroutine) and of an arguments
list.

This method will actually fork a new process and execute the given code
reference in the child process. For the parent process this method will return
automatically. The child process will start executing the code reference with
the given arguments.

The parent process, the one that started the task should wait for the child
process to resume. This can be performed individually on each tasks through the
method L<"Parallel::SubFork::Task/wait_for"> or for all tasks launched through
this instance through the method L<"wait_for_all">

B<NOTE:> This method requires that the caller process is the same process as the
one that created the instance object being called.

Parameters:

=over

=item $code

The code reference to execute.

=item @args (optional)

The arguments to pass to the code reference.

=back

=cut

sub start {

	# Arguments
	my $self = shift;
	my ($code, @args) = @_;

	# Stop if this is not the dispatcher
	$self->_assert_is_dispatcher();


	# Start the task and remember it
	my $task;
	eval {
		$task = Parallel::SubFork::Task->start($code, @args);
	};
	if (my $error = $@) {
		croak $error;
	}
	push @{ $self->{tasks} }, $task;
	
	return $task;
}


=head2 wait_for_all

This method waits for all tasks started so far and returns when they all have
resumed. This is useful for creating a rally point for multiple tasks.

B<NOTE:> This method requires that the caller process is the same process as the
one that created the instance object being called.

=cut

sub wait_for_all {
	my $self = shift;

	$self->_assert_is_dispatcher();

	foreach my $task ($self->tasks) {
		eval {
			$task->wait_for();
		};
		if (my $error = $@) {
			croak $error;
		}
	}
}


=head2 tasks

Returns the tasks started so far by this instance. This method returns a list
and not an array ref.

=cut

sub tasks {
	my $self = shift;

	my $tasks = $self->{tasks};
	my @tasks = defined $tasks ? @{ $tasks } : ();
	return @tasks;
}


=head2 _assert_is_dispatcher

Used to check if the current process is the same one that invoked the
constructor.

This is required as only the dispatcher process is allowed to start and wait for
tasks.

=cut

sub _assert_is_dispatcher {
	my $self = shift;
	return if $self->_dispatcher_pid == $$;
	croak "Process $$ is not the main dispatcher";
}


# Return a true value
1;


=head1 NOTES

The API is not yet frozen and could change as the module goes public.

=head1 SEE ALSO

Take a look at L<POE> for asynchronous multitasking and networking.

=head1 AUTHOR

Emmanuel Rodriguez, E<lt>emmanuel.rodriguez@gmail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2008-2010 by Emmanuel Rodriguez

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.

=cut


lib/Parallel/SubFork/Task.pm
package Parallel::SubFork::Task;

=head1 NAME

Parallel::SubFork::Task - Run Perl functions in forked processes. 

=head1 SYNOPSIS

	use Parallel::SubFork::Task;
	
	# Run a some arbitrary Perl code in a separated process
	my $task = Parallel::SubFork::Task->start(\&job, @args);
	$task->wait_for();
	
	# Create and execute the task (same as above)
	my $task2 = Parallel::SubFork::Task->new(\&job, @args);
	$task2->execute();
	$task2->wait_for();
	
	# Wait with a live progress
	local $| = 1; # Force print to flush the output
	my $task3 = Parallel::SubFork::Task->new(\&job, @args);
	while ($task3->wait_for(0.5)) {
		print ".";	
	}
	
	# Access any of the properties
	printf "PID of task was %s\n", $task->pid;
	printf "Args of task where %s\n", join(", ", $task->args);
	printf "Exit code: %d\n", $task->exit_code;

=head1 DESCRIPTION

This module provides a simpler way to run arbitrary Perl code in a different
process. This module consists of a fancy wrapper over the system calls C<fork>
and C<waitpid>. The idea is to execute any standard Perl function in a different
process without any of the inconveniences of managing the forks by hand.

=head1 TASK

This module is used to encapsulate a task, i.e. the function to be executed in
a different process and it's arguments. In a nutshell a task consists of a
reference to a Perl function (C<\&my_sub>) or a closure (C<sub { 1; }>), 
also known as an anonymous subroutine, and optionally the arguments to provide
to that function.

A task also stores some runtime properties such as the PID of the process that 
executed the code, the exit code and the exit status of the process. These
properties can then be inspected by the parent process through their dedicated
accessors.

There's also some helper methods that are used to create the child process and
to wait for it to resume.

=head1 PROCESSES

Keep in mind that the function being executed is run in a different process.
This means that any modification performed within that function will only affect
the process running the task. This is true even for global variables. All data
exchange or communication between the parent the child process has to be
implemented manually through standard I<inter process communication> (IPC)
mechanisms (see L<perlipc>).

The child process used to executes the Perl subroutines has it's environment
left unchanged. This means that all file descriptors, signal handlers and other
resources are still available. It's up to the subroutine to prepare it self a
proper environment.

=head1 RETURN VALUES

The subroutine return's value will be used as the process exit code, this is the
only thing that the invoking process will be able to get back from the task
without any kind of IPC. This means that the return value should be an integer.
Furthermore, since the return value is used as an exit value in this case C<0>
is considered as successful execution while any other value is usually
interpreted as an error.

=head1 EXIT

The subroutine is free to raise any exceptions through C<die> or any similar
mechanism. If an error is caught by the framework it will be interpreted as an
error and an appropriate exit value will be used.

If the subroutine needs to resume it's execution through a the system call
C<exit> then consider instead using C<_exit> as defined in the module L<POSIX>.
This is because C<exit> not only terminates the current process but it performs
some cleanup such as calling the functions registered with C<atexit> and flush
all stdio streams before finishing the process. Normally, only the main process
should call C<exit>, in the case of a fork the children should finish their
execution through C<POSIX::_exit>.

=head1 PROCESS WAIT

Waiting for process to finish can be problematic as there are multiple ways for
waiting for processes to resume each having it's advantages and disadvantages.

The easiest way is to register a signal handler for C<CHLD> signal. This has the
advantage of receiving the child notifications as they happen, the disadvantage
is that there's no way to control for which children the notifications will
happen. This is quite inconvenient because a lot of the nice built-in functions
and operators in Perl such as C<`ls`>, C<system> and even C<open> (when used in
conjunction with a C<|>) use child processes for their tasks and this could
potentially interfere with such utilities.

Another alternative is to wait for all processes launched but this can also
interfere with other processed launched manually through C<fork>.

Finally, the safest way is to wait explicitly B<only> for the processes that we
know to have started and nothing else. This there will be no interference with
the other processes. This is exactly the approach used by this module.

=head1 METHODS

A task defines the following methods:

=cut


use strict;
use warnings;

use POSIX qw(
	WNOHANG
	WIFEXITED
	WEXITSTATUS
	WIFSIGNALED
	_exit
);

use Carp;

use base qw(Class::Accessor::Fast);
__PACKAGE__->mk_accessors(
	qw(
		_ppid
		pid
		code
		exit_code
		status
	)
);


# Version of the module
our $VERSION = '0.08';


# Check if it's possible to use a high precision alarm
my $HIRES; # NOTE the initialization must be done in the BEGIN block otherwise
           #      the default value will override whatever was set in the BEGIN
					 #      block.
BEGIN {
	$HIRES = 0; # Assume that there's no HiRes
	eval {
		require Time::HiRes;
		$HIRES = 1;
	};
}


=head2 start

Creates and executes a new task, this is simply a small shortcut for starting
new tasks.

In order to manage tasks easily consider using use the module
L<Parallel::SubFork> instead.

Parameters:

	$code: the code reference to execute in a different process.
	@args: the arguments to pass to the code reference (optional).

=cut

sub start {
	my $class = shift;
	my ($code, @args) = @_;
	croak "First parameter must be a code reference" unless ref $code eq 'CODE';
	
	my $task = $class->new($code, @args);
	$task->execute();

	return $task;
}


=head2 new

Creates a new task, this is simply a constructor and the task will not be
started yet.

The task can latter by started through a call to L</"execute">.

In order to manage tasks easily consider using use the module
L<Parallel::SubFork> instead.

Parameters:

=over

=item $code

The code reference to execute.

=item @args (optional)

The arguments to pass to the code reference.

=back

=cut

sub new {
	my $class = shift;
	my ($code, @args) = @_;
	croak "First parameter must be a code reference" unless ref $code eq 'CODE';
	
	# Create a blessed instance
	my $self = bless {}, ref($class) || $class;
	$self->code($code);
	$self->{args} = \@args;
	
	return $self;
}


=head2 code

Accessor to the function (code reference) that will be executed in a different
process. This is what the child process will execute. 

This function is expected to return C<0> for success and any other integer to
indicate a failure. The function is free to raise any kind of exception as the
framework will catch all exceptions and return an error value instead.

The function will receive it's parameters normally through the variable C<@_>.

=head2 pid

The PID of the process executing the subroutine, the child's PID.

=head2 exit_code

The exit code of the task, this is the value returned by C<exit>, 
C<POSIX::_exit> or C<return>.

=head2 status

The exit code returned to the parent process as described by C<wait>. The status
code can be inspected through the L<"POSIX/WAIT"> macros .

=head2 args

The arguments that will be given to the subroutine being executed in a separated
process. The subroutine will receive this very same arguments through C<@_>.

This method always return it's values as a list and not as an array ref.

=cut

sub args {
	my $self = shift;
	
	my $args = $self->{args};
	my @args = defined $args ? @{ $args } : ();
	return @args;
}


=head2 execute

Executes the tasks (the code reference encapsulated by this task) in a new
process. The code reference will be invoked with the arguments passed in the
constructor.

This method performs the actual fork and returns automatically for the invoker,
while the child process will start to execute the code in defined in the code
reference. Once the subroutine has finished the child process will resume right
away.

The invoker (the parent process) should call L</wait_for> in order to wait for
the child process to finish and obtain it's exit value.

=cut

sub execute {
	my $self = shift;

	# Check that we don't run twice the same task
	if (defined $self->pid) {
		croak "Task already exectuted";
	}
	
	# Make sure that there's a code reference
	my $code = $self->code;
	if (! (defined $code and ref $code eq 'CODE')) {
		croak "Task requires a valid code reference (function)";
	}

	my $ppid = $$;

	# Fork a child
	my $pid = fork();
	
	# Check if the fork succeeded
	if (! defined $pid) {
		croak "Can't fork because: $!";
	}
	
	$self->_ppid($ppid);
	if ($pid == 0) {
		## CHILD part

		# Execute the main code
		my $return = 1;
		eval {
			$return = $code->($self->args);
		};
		if (my $error = $@) {
			carp "Child executed with errors: ", $error;
		}
		
		# This is as far as the kid gets if the callback hasn't called exit we do it
		_exit($return);
	}
	else {
		## PARENT part
		$self->pid($pid);
	}
}


=head2 wait_for

Waits until the process running the task (the code reference) has finished. By
default this method waits forever until task resumes either naturally or due to
an error.

If a parameter is passed then it is assumed to be the number of seconds to wait.
Once the timeout has expired the method will return with a true value. This is
the only condition under which the method will return with a true value.

If the module L<Time::HiRes> is available then timeout can be in fractions (ex:
0.5 for half a second) otherwise full integers have to be provided. If not Perl
will round the results during the conversion to int.

The timeout is implemented through C<sleep> and has all the caveats of sleep,
see perdoc -f sleep for more details. Remember that sleep could take a second
less than requested (sleep 1 could do no sleep at all) and mixin calls to sleep
and alarm is at your own risks as sleep is sometimes implemented through alarm.
Furthermore, if a timeout between 0 and 1 second is provided as a fraction and
that C<Time::Hires> is not available Perl will round the value to 0.

The exit status of the process can be inspected through the accessor 
L</"exit_code"> and the actual status, the value returned in C<$?> by C<waitpid>
can be accessed through the accessor L</"status">.

Parameters:

=over

=item $timeout (optional)

The number of seconds to wait until the method returns due to a timeout. If
undef then the method doesn't apply a timeout and waits until the task has
resumed.

=back

Returns:

If the method was invoked without a timeout then a false value will always be
returned, no matter the outcome of the task. If a timeout was provided then the
method will return a true value only when the timeout has been reached otherwise
a false value will be returned.

=cut

sub wait_for {
	my $self = shift;
	my ($timeout) = @_;

	my $pid = $self->pid;
	return unless defined $pid and $pid > 0;
	if (! (defined $pid and $pid > 0) ) {
		croak "Task isn't started";
	}
	
	# Only the real parent can wait for the child
	if ($self->_ppid != $$) {
		croak "Only the parent process can wait for the task";
	}
	
	# Check if the task was already waited for
	if (defined $self->status) {
		return;
	}
	
	my $timemout_done = 0; # Use to track if the waitpid was called enough times when passed a timeout
	my $flags = defined $timeout ? WNOHANG : 0;
	while (1) {
		
		# Wait for the specific PID
		my $result = waitpid($pid, $flags);

		if ($result == -1) {
			# No more processes to wait for, but we didn't find our PID
			croak "No more processes to wait PID $pid not found";
		}
		elsif ($result == 0) {
			# The process is still running

			# If the method was called with a timeout we will retry waitpid once more;
			# remember that it is invoked with no hang which means that the call will
			# return instantaneously.
			if (defined $timeout) {

				# In the case of a timeout we invoke this code once
				$timemout_done++ or return 1;

				# NOTE: The timeout is implemented with a sleep instead of an alarm
				#       because some versions/combinations of perl and Time::HiRes cause
				#       Time::HiRes::alarm() to fail to interrupt system calls. For more
				#       information about this see Ticket #51465:
				#          https://rt.cpan.org/Ticket/Display.html?id=51465

				# Sleep and alarms don't mix well together, so we stop the current alarm
				# and restore it later on.
				my $alarm = alarm(0);
				if ($HIRES) {
					Time::HiRes::sleep($timeout);
				}
				else {
					sleep($timeout);
				}

				# If an alarm was set, restore it
				alarm($alarm) if $alarm;
			}

			# Continue waiting as the process is till waiting
			next;
		}
		elsif ($result != $pid) {
			# Strange we got another PID than ours
			croak "Got a status change for PID $result while waiting for PID $pid";
		}
		
		# Now we got a decent answer from waitpid, this doesn't mean that the child
		# died! It just means that the child got a state change (the child
		# terminated; the child was stopped by a signal; or  the  child was  resumed
		# by a signal). Here we must check if the process finished properly
		# otherwise we must continue waiting for the end of the process.
		my $status = $?;
		if (WIFEXITED($status)) {
			$self->status($status);
			$self->exit_code(WEXITSTATUS($status));
			return;
		}
		elsif (WIFSIGNALED($status)) {
			$self->status($status);
			# WEXITSTATUS is only defined for WIFEXITED, here we assume an error
			$self->exit_code(1);
			return;
		}
	}
	
	return;
}


=head2 kill

Sends a signal to the process. This is a simple wrapper over the system call
C<kill>. It takes the kind of signal that the built-in kill function.

B<NOTE>: Calling kill doesn't warranty that the task will die. Most signals can
be caught by the process and may not kill it. In order to be sure that the
process is killed it is advised to call L</wait_for>. Even if the signal kills
the process L</wait_for> has to be called otherwise the task's process will be
flagged as zombie process (see L<http://en.wikipedia.org/wiki/Zombie_process>).

The following code snippet shows how to properly kill a task:

	my $task = Parallel::SubFork::Task->start(\&job);
	if ($task->wait_for(2)) {
		# Impatient block
		$task->kill('KILL');
		$task->wait_for();
	}

Parameters:

=over

=item $signal

The signal to send to the process. Same as the first parameter passed to the
Perl built-in.

=back

Returns:

The same value as Perl's C<kill>.

=cut

sub kill {
	my $self = shift;
	my ($signal) = @_;
	kill $signal, $self->pid;
}


# Return a true value
1;

=head1 NOTES

The API is not yet frozen and could change as the module goes public.

=head1 SEE ALSO

Take a look at L<POE> for asynchronous multitasking and networking.

=head1 AUTHOR

Emmanuel Rodriguez, E<lt>emmanuel.rodriguez@gmail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2008-2010 by Emmanuel Rodriguez

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.

=cut
